File: ios/Classes/AmplitudeExperimentPlugin.swift
import AmplitudeExperiment
import Flutter
import UIKit

public class AmplitudeExperimentPlugin: NSObject, FlutterPlugin, AmplitudeExperimentHostApi {
    private var instances: [String: ExperimentClient] = [:]
    var providerApi: CustomProviderApi!

    public static func register(with registrar: FlutterPluginRegistrar) {
        let instance = AmplitudeExperimentPlugin()
        let messenger = registrar.messenger()
        AmplitudeExperimentHostApiSetup.setUp(binaryMessenger: messenger, api: instance)
        instance.providerApi = CustomProviderApi(binaryMessenger: messenger)
    }

    private func requireClient(_ instanceName: String) throws -> ExperimentClient {
        guard let client = instances[instanceName] else {
            throw PigeonError(
                code: "INSTANCE_NOT_FOUND",
                message: "Experiment instance \(instanceName) not found",
                details: nil
            )
        }
        return client
    }

    func initializeExperiment(apiKey: String, config: ExperimentConfigData) throws {
        let sdkConfig = ExperimentSdkCodec.convertConfig(config, api: providerApi)
        let client = Experiment.initialize(apiKey: apiKey, config: sdkConfig)
        instances[config.instanceName] = client
    }

    func initializeExperimentWithAmplitude(apiKey: String, config: ExperimentConfigData) throws {
        let sdkConfig = ExperimentSdkCodec.convertConfig(config, api: providerApi)
        let client = Experiment.initializeWithAmplitudeAnalytics(apiKey: apiKey, config: sdkConfig)
        instances[config.instanceName] = client
    }

    func start(instanceName: String, user: ExperimentUser?, completion: @escaping (Result<Void, Error>) -> Void) {
        do {
            let client = try requireClient(instanceName)
            let sdkUser = ExperimentSdkCodec.convertUser(user)
            client.start(sdkUser) { error in
                if let error = error {
                    completion(.failure(error))
                } else {
                    completion(.success(()))
                }
            }
        } catch {
            completion(.failure(error))
        }
    }

    func stop(instanceName: String) throws {
        let client = try requireClient(instanceName)
        client.stop()
    }

    func fetch(instanceName: String, user: ExperimentUser?, options: FetchOptions?, completion: @escaping (Result<Void, Error>) -> Void) {
        do {
            let client = try requireClient(instanceName)
            let sdkUser = ExperimentSdkCodec.convertUser(user)
            let sdkOptions = ExperimentSdkCodec.convertFetchOptions(options)
            client.fetch(user: sdkUser, options: sdkOptions) { _, error in
                if let error = error {
                    completion(.failure(error))
                } else {
                    completion(.success(()))
                }
            }
        } catch {
            completion(.failure(error))
        }
    }

    func variant(instanceName: String, user: ExperimentUser, flagKey: String, fallbackVariant: Variant?) throws -> Variant {
        let client = try requireClient(instanceName)
        if let sdkUser = ExperimentSdkCodec.convertUser(user) {
            client.setUser(sdkUser)
        }
        let fallback = ExperimentSdkCodec.convertVariant(fallbackVariant)
        let sdkVariant = client.variant(flagKey, fallback: fallback)
        return ExperimentSdkCodec.convertVariant(sdkVariant)
    }

    func all(instanceName: String, user: ExperimentUser) throws -> [String: Variant] {
        let client = try requireClient(instanceName)
        if let sdkUser = ExperimentSdkCodec.convertUser(user) {
            client.setUser(sdkUser)
        }
        let allVariants = client.all()
        return allVariants.mapValues { ExperimentSdkCodec.convertVariant($0) }
    }

    func clear(instanceName: String) throws {
        let client = try requireClient(instanceName)
        client.clear()
    }

    func exposure(instanceName: String, key: String) throws {
        let client = try requireClient(instanceName)
        client.exposure(key: key)
    }

    func getUser(instanceName: String) throws -> ExperimentUser {
        let client = try requireClient(instanceName)
        let sdkUser = client.getUser()
        return ExperimentSdkCodec.convertUserToPigeon(sdkUser)
    }

    func setUser(instanceName: String, user: ExperimentUser) throws {
        let client = try requireClient(instanceName)
        guard let sdkUser = ExperimentSdkCodec.convertUser(user) else {
            throw PigeonError(
                code: "INVALID_ARGUMENT",
                message: "user is required",
                details: nil
            )
        }
        client.setUser(sdkUser)
    }

    func setTracksAssignment(instanceName: String, tracksAssignment: Bool) throws {
        let client = try requireClient(instanceName)
        client.setTracksAssignment(tracksAssignment)
    }
}

File: android/src/main/kotlin/com/amplitude/experiment/flutter/AmplitudeExperimentPlugin.kt
package com.amplitude.experiment.flutter

import android.app.Application
import android.content.Context
import com.amplitude.experiment.Experiment
import com.amplitude.experiment.ExperimentClient
import io.flutter.embedding.engine.plugins.FlutterPlugin
import java.util.concurrent.ConcurrentHashMap

class AmplitudeExperimentPlugin :
    FlutterPlugin,
    AmplitudeExperimentHostApi {
    private lateinit var ctxt: Context
    private var instances:  Map<String, ExperimentClient> = ConcurrentHashMap()
    private lateinit var providerApi: CustomProviderApi

    private val executor = java.util.concurrent.Executors.newCachedThreadPool()
    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        ctxt = flutterPluginBinding.applicationContext
        AmplitudeExperimentHostApi.setUp(flutterPluginBinding.binaryMessenger, this)
        providerApi = CustomProviderApi(flutterPluginBinding.binaryMessenger)
    }

    override fun onDetachedFromEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        AmplitudeExperimentHostApi.setUp(flutterPluginBinding.binaryMessenger, null)
        executor.shutdown()
    }

    override fun init(
        apiKey: String,
        config: ExperimentConfigData
    ) {
        if (instances.contains(config.instanceName)) {
            return
        }

        val client = Experiment.initialize(ctxt as Application, apiKey, convertConfig(config, providerApi))
        instances += mapOf(config.instanceName to client)
    }

    override fun initWithAmplitude(
        apiKey: String,
        config: ExperimentConfigData
    ) {
        if (instances.contains(config.instanceName)) {
            return
        }

        val client = Experiment.initializeWithAmplitudeAnalytics(ctxt as Application, apiKey, convertConfig(config, providerApi))
        instances += mapOf(config.instanceName to client)
    }

    override fun start(
        instanceName: String,
        user: ExperimentUser?,
        callback: (Result<Unit>) -> Unit
    ) {
        executor.execute {
            try {
                getClient(instanceName).start(convertUser(user))
                callback(Result.success(Unit))
            } catch (e: Exception) {
                callback(Result.failure(e))
            }
        }
    }

    override fun stop(instanceName: String) {
        getClient(instanceName).stop()
    }

    override fun fetch(
        instanceName: String,
        user: ExperimentUser?,
        options: FetchOptions?,
        callback: (Result<Unit>) -> Unit
    ) {
        executor.execute {
            try {
                getClient(instanceName).fetch(convertUser(user), convertFetchOptions(options)).get()
                callback(Result.success(Unit))
            } catch (e: Exception) {
               callback(Result.failure(e))
            }
        }
    }

    override fun variant(
        instanceName: String,
        user: ExperimentUser?,
        flagKey: String,
        fallbackVariant: Variant?
    ): Variant {
        val client = getClient(instanceName)
        convertUser(user)?.let { client.setUser(it) }
        return variantToPigeon(client.variant(flagKey, variantFromPigeon(fallbackVariant)))
    }

    override fun all(instanceName: String, user: ExperimentUser?): Map<String, Variant> {
        val client = getClient(instanceName)
        convertUser(user)?.let { client.setUser(it) }
        return variantsToPigeon(client.all())
    }

    override fun clear(instanceName: String) {
        getClient(instanceName).clear()
    }

    override fun exposure(instanceName: String, key: String) {
        getClient(instanceName).exposure(key)
    }

    override fun getUser(instanceName: String): ExperimentUser {
        val sdkUser = requireNotNull(getClient(instanceName).getUser()) { "native getUser() returned null for instance $instanceName" }
        return convertUser(sdkUser)
    }

    override fun setUser(
        instanceName: String,
        user: ExperimentUser
    ) {
        val sdkUser = requireNotNull(convertUser(user)) { "user conversion must not return null for non-null input" }
        getClient(instanceName).setUser(sdkUser)
    }

    override fun setTracksAssignment(
        instanceName: String,
        tracksAssignment: Boolean
    ) {
        getClient(instanceName).setTracksAssignment(tracksAssignment)
    }

    private fun getClient(instanceName: String): ExperimentClient {
        return instances[instanceName] ?: throw IllegalArgumentException("Instance $instanceName has not been initialized")
    }
}
